### ISSUE TITLE
TypeError: Cannot redefine property occurs when using MUI v5 and shadow-cljs 2.17.0+

### ISSUE THREAD

**Issue by arttuka** [thheller/shadow-cljs#992]
The same error was already discussed in #970, and it didn't happen in shadow-cljs v2.16.*, but it has reappeared in v2.17.0. I decided to open a new issue since I have some more information related to the error.

The issue happens due to MUI package structure with both CJS and ESM modules, where the CJS modules are built by Babel. The structure looks like this:
```
@mui/material/index.js             # "@mui/material" as ESM
@mui/material/package.json         # package.json for "@mui/material"
@mui/material/Button/index.js      # "@mui/material/Button" as ESM
@mui/material/Button/package.json  # package.json for "@mui/material/Button"
@mui/material/node/index.js        # "@mui/material" as CJS
@mui/material/node/Button/index.js # "@mui/material/Button" as CJS
```

`@mui/material/package.json` defines
```
  "main": "./node/index.js",
  "module": "./index.js",
```

`@mui/material/Button/package.json` defines
```
  "main": "../node/Button/index.js",
  "module": "./index.js",
```

By default, shadow-cljs loads `main` sources before `module` sources, so requiring `"@mui/material"` loads `@mui/material/node/index.js`. The problem is that requiring `"@mui/material/Button"` loads `@mui/material/Button/index.js` instead of `@mui/material/node/Button/index.js`, even though `@mui/material/Button/package.json` defines the correct `main`.

The following chain of requires then happens:
```
@mui/material (package require, CJS)
-> @mui/base (package require, CJS)
  -> @mui/base/ModalUnstyled (relative require, CJS)
    -> @mui/base/ModalUnstyled/ModalManager (relative require, CJS)
-> @mui/material/Modal (relative require, CJS)
  -> @mui/base/ModalUnstyled (package require, ESM)                 <- this is where the wrong file is loaded
    -> @mui/base/ModalUnstyled/ModalManager (relative require, ESM)
```
Now, `@mui/material` has required `@mui/base/ModalUnstyled/ModalManager` as both CJS and ESM, and is trying to export both. If both requires were CJS, this would be fine since Babel's re-export checks for equality first. In this case, they aren't equal, they are being exported with the same name, and this results in `TypeError: Cannot redefine property: ModalManager`. This only happens when one JS file tries to export the same component with the same name via two paths, that result in CJS and ESM files being required.

It seems that shadow-cljs 2.17.0 ignores the package.json in subdirectories of a package. I have made a reproduction of this issue at https://github.com/arttuka/shadow-cljs-issue-reproduction . It contains both the above issue, as well as a simple reproduction using dummy JS libraries. 

Even in cases where this bug doesn't cause an outright error, it can cause multiple copies of the same file to be included in the built JS, increasing bundle size.

Setting `:js-options {:entry-keys ["module" "browser" "main"]}` solves this issue in some cases, but causes other errors in many cases (as shadow-cljs user guide points out).

**Comment by arttuka** [comment]
I found some new information and edited the description. It seems that shadow-cljs 2.17.0 ignores the package.json in subdirectories of a package.

**Comment by thheller** [comment]
Thanks for the detailed report/repro. Fixed in `2.17.5`.
